Integer Overflow in Fair Scheduler on Priority Change

### Describe the bug
The bug is triggered by a race condition where a newly created process has its priority and scheduling weight drastically increased. The kernel fails to update the scheduler's total run queue weight to match the process's new, larger weight. When the scheduler later tries to account for this process, it subtracts the large new weight from the small old total, causing an integer underflow and a kernel panic.

https://github.com/asterinas/asterinas/blob/780b35848e484bb091725c232a8c4ba09d5d7765/kernel/src/sched/sched_class/fair.rs#L246-L253

### To Reproduce
1. Compile the program and run:
```C
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/resource.h>
#include <sched.h>
#include <signal.h>
#include <string.h>
#include <sys/syscall.h>
#include <linux/types.h>

int child_func(void *arg) {
    while (1) {}
    return 0;
}

int main() {
    char *stack = malloc(8192);
    if (!stack) {
        perror("malloc");
        exit(1);
    }
    char *stack_top = stack + 8192;

    pid_t pid = clone(child_func, stack_top, SIGCHLD, NULL);

    if (pid == -1) {
        perror("clone failed");
        free(stack);
        exit(1);
    }

    usleep(100000);

    struct sched_attr {
        __u32 size;
        __u32 sched_policy;
        __u64 sched_flags;
        __s32 sched_nice;
        __u32 sched_priority;
        __u64 sched_runtime;
        __u64 sched_deadline;
        __u64 sched_period;
    } attr;

    memset(&attr, 0, sizeof(attr));
    attr.size = sizeof(attr);
    attr.sched_policy = SCHED_OTHER;
    attr.sched_nice = -20;

    if (syscall(SYS_sched_setattr, pid, &attr, 0) == -1) {
        perror("sched_setattr failed");
        kill(pid, SIGKILL);
        waitpid(pid, NULL, 0);
        free(stack);
        exit(1);
    }

    printf("Parent: Priority changed. Waiting for kernel panic...\n");
    waitpid(pid, NULL, 0);

    free(stack);
    return 0;
}
```

### Expected behavior
The Asterinas should not panic.

### Environment
- Official docker asterinas/asterinas:0.15.0-20250603
- Intel(R) Xeon(R) Gold 6230R CPU @ 2.10GHz
- Asterinas version: main 780b358

### Log
```
~ # /root/fair.c 
Parent: Priority changed. Waiting for kernel panic...
[   200.592] ERROR: Uncaught panic:
        attempt to subtract with overflow
        at /root/asterinas/kernel/src/sched/sched_class/fair.rs:250
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicCpuSet { bits: [1] }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31486020 }, fair: FairAttr { weight: 1024, vruntime: 396319599 } } })
Printing stack trace:
   1: fn 0xffffffff8881b450 - pc 0xffffffff8881b49a / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedaf690;

   2: fn 0xffffffff88091170 - pc 0xffffffff88091919 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedaf6e0;

   3: fn 0xffffffff88091160 - pc 0xffffffff8809116a / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafaf0;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafb00;

   5: fn 0xffffffff88951280 - pc 0xffffffff889512c4 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafb10;

   6: fn 0xffffffff88933540 - pc 0xffffffff8893356e / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafb50;

   7: fn 0xffffffff88502970 - pc 0xffffffff88502a88 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafba0;

   8: fn 0xffffffff880b3650 - pc 0xffffffff880b365e / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafc20;

   9: fn 0xffffffff8828e3c0 - pc 0xffffffff8828e410 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafc30;

  10: fn 0xffffffff8818bf60 - pc 0xffffffff8818bf9b / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafc70;

  11: fn 0xffffffff8818c1c0 - pc 0xffffffff8818c1d3 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafc90;

  12: fn 0xffffffff8877b550 - pc 0xffffffff8877b587 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafcb0;

  13: fn 0xffffffff8877bae0 - pc 0xffffffff8877bb01 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafd30;

  14: fn 0xffffffff8818ba70 - pc 0xffffffff8818bb08 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafd70;

  15: fn 0xffffffff8877b810 - pc 0xffffffff8877b88b / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafdd0;

  16: fn 0xffffffff8877b500 - pc 0xffffffff8877b53f / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafe30;

  17: fn 0xffffffff88772180 - pc 0xffffffff887721c4 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafe70;

  18: fn 0xffffffff887720d0 - pc 0xffffffff887720f0 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafea0;

  19: fn 0xffffffff8841e5b0 - pc 0xffffffff8841e70d / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedafec0;

  20: fn 0xffffffff8842dfe0 - pc 0xffffffff8842e07e / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedaff80;

  21: fn 0xffffffff8842b6d0 - pc 0xffffffff8842b798 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0030;

  22: fn 0xffffffff8842c2b0 - pc 0xffffffff8842c3d3 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb00c0;

  23: fn 0xffffffff8842b5a0 - pc 0xffffffff8842b5c6 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0170;

  24: fn 0xffffffff885a5b20 - pc 0xffffffff885a5b73 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0190;

  25: fn 0xffffffff885021f0 - pc 0xffffffff885022a8 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb01d0;

  26: fn 0xffffffff880b6610 - pc 0xffffffff880b66ad / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0270;

  27: fn 0xffffffff880d2640 - pc 0xffffffff880d2963 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0410;

  28: fn 0xffffffff8844c650 - pc 0xffffffff8845b240 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedb0960;

  29: fn 0xffffffff8844be40 - pc 0xffffffff8844bed1 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2640;

  30: fn 0xffffffff8813a8b0 - pc 0xffffffff8813b0f4 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc27c0;

  31: fn 0xffffffff8823c190 - pc 0xffffffff8823c196 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2d00;

  32: fn 0xffffffff880cb230 - pc 0xffffffff880cb247 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2d10;

  33: fn 0xffffffff880cb130 - pc 0xffffffff880cb151 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2d30;

  34: fn 0xffffffff884ad280 - pc 0xffffffff884ad292 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2d80;

  35: fn 0xffffffff88226b00 - pc 0xffffffff88226b16 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2db0;

  36: fn 0xffffffff88202dd0 - pc 0xffffffff88202ddc / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2f50;

  37: fn 0xffffffff8823b560 - pc 0xffffffff8823b56e / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2f80;

  38: fn 0xffffffff887e6040 - pc 0xffffffff887e6065 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2fa0;

  39: fn 0xffffffff8883abe0 - pc 0xffffffff8883ac79 / registers:

     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc2fd0;


     rax 0xffffffff88c91010; rdx  0x100000000000008; rcx 0xffffffff88c91010; rbx                0x0;
     rsi 0xffffffff88c91011; rdi  0x1ffffff887fb5d7; rbp                0x0; rsp 0xffffdffffedc3000;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panic.rs:107
(  2) /root/asterinas/kernel/src/thread/oops.rs:129
(  3) ??:?
(  4) bg4c9y98l0p8tlbbqav3ajkcf:?
(  5) ??:?
(  6) ??:?
(  7) /root/asterinas/kernel/src/sched/sched_class/fair.rs:253
(  8) /root/asterinas/kernel/src/sched/sched_class/mod.rs:303
(  9) /root/.rustup/toolchains/nightly-2025-02-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1552
( 10) /root/asterinas/kernel/src/sched/sched_class/mod.rs:301
( 11) /root/asterinas/kernel/src/sched/sched_class/mod.rs:333
( 12) /root/asterinas/ostd/src/task/scheduler/mod.rs:147
( 13) /root/asterinas/ostd/src/task/scheduler/mod.rs:246
( 14) /root/asterinas/kernel/src/sched/sched_class/mod.rs:233
( 15) /root/asterinas/ostd/src/task/scheduler/mod.rs:249
( 16) /root/asterinas/ostd/src/task/scheduler/mod.rs:128
( 17) /root/asterinas/ostd/src/sync/wait.rs:274
( 18) /root/asterinas/ostd/src/sync/wait.rs:201
( 19) /root/asterinas/ostd/src/sync/wait.rs:229
( 20) /root/asterinas/kernel/src/time/wait.rs:210
( 21) /root/asterinas/kernel/src/process/signal/pause.rs:133
( 22) /root/asterinas/kernel/src/process/signal/pause.rs:201
( 23) ??:?
( 24) /root/asterinas/kernel/src/process/wait.rs:70
( 25) /root/asterinas/kernel/src/process/signal/pause.rs:222
( 26) /root/asterinas/kernel/src/process/wait.rs:66
( 27) /root/asterinas/kernel/src/syscall/wait4.rs:26
( 28) /root/asterinas/kernel/src/syscall/mod.rs:202
( 29) /root/asterinas/kernel/src/syscall/mod.rs:352
( 30) /root/asterinas/kernel/src/thread/task.rs:87
( 31) /root/.rustup/toolchains/nightly-2025-02-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 32) /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.5/src/panicking.rs:56
( 33) /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.5/src/panicking.rs:42
( 34) /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.5/src/panic.rs:87
( 35) /root/asterinas/kernel/src/thread/oops.rs:54
( 36) /root/asterinas/kernel/src/thread/task.rs:117
( 37) /root/.rustup/toolchains/nightly-2025-02-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 38) /root/.rustup/toolchains/nightly-2025-02-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2016
( 39) /root/asterinas/ostd/src/task/mod.rs:234
make: *** [Makefile:238: run] Error 1
```

--- a/kernel/src/sched/sched_class/fair.rs
+++ b/kernel/src/sched/sched_class/fair.rs
@@ -3,7 +3,7 @@
 use alloc::{collections::BinaryHeap, sync::Arc};
 use core::{
     cmp::{self, Reverse},
-    sync::atomic::{AtomicU64, Ordering::Relaxed},
+    sync::atomic::{AtomicU64, Ordering},
 };
 
 use ostd::{
@@ -24,6 +24,9 @@ use crate::{
 };
 
 const WEIGHT_0: u64 = 1024;
+
+const HAS_PENDING: u64 = 1 << (u64::BITS - 1);
+
 pub const fn nice_to_weight(nice: Nice) -> u64 {
     // Calculated by the formula below:
     //
@@ -53,6 +56,7 @@ pub const fn nice_to_weight(nice: Nice) -> u64 {
                     WEIGHT_0 * numerator / denominator
                 }
             };
+            assert!(ret[index] & HAS_PENDING == 0);
 
             index += 1;
             nice += 1;
@@ -93,9 +97,38 @@ pub const fn nice_to_weight(nice: Nice) -> u64 {
 ///
 ///     period_delta > time_slice
 ///         || vruntime > rq_min_vruntime + normalized_time_slice
+///
+/// # The weight update process
+///
+/// The weight of a thread can be updated by the `sched_setattr` syscall series in
+/// any thread. This makes it difficult to re-evaluate the data of its run queue
+/// instantly after the update without a direct backward reference (which is
+/// impossible to be represented in safe Rust).
+///
+/// To handle this problem, we use a `pending_weight` field to store the new weight.
+/// When the thread is scheduled within the run queue, we will check if the weight
+/// needs to be updated since both the old and new weights are needed for re-evaluation.
+///
+/// To indicate whether the weight needs to be updated, we pack the `weight` field
+/// with a bit flag `HAS_PENDING`. The overall mechanism is similar to an optimized
+/// version of spin locks. When accessing the `weight` field:
+///
+/// - If the weight does not need to be updated (i.e. `weight & IS_PENDING == 0`),
+///   we simply return the weight.
+/// - If the weight needs to be updated (i.e. `weight & IS_PENDING != 0`), we try to
+///   store the new weight into the `weight` field with `IS_PENDING` cleared via a
+///   `compare_exchange_weak` loop, which shouldn't take too much time since the update
+///   frequency is usually relatively low.
+/// - If the result of the loop turns out that the weight doesn't need to be updated, we
+///   return the weight directly.
+/// - After a successful update, we re-evaluate the data of the run queue.
+///
+/// This method allows the access to the weight lock-free and ensures only 1 load
+/// is needed most of the time.
 #[derive(Debug)]
 pub struct FairAttr {
     weight: AtomicU64,
+    pending_weight: AtomicU64,
     vruntime: AtomicU64,
 }
 
@@ -103,19 +136,53 @@ impl FairAttr {
     pub fn new(nice: Nice) -> Self {
         FairAttr {
             weight: nice_to_weight(nice).into(),
+            pending_weight: Default::default(),
             vruntime: Default::default(),
         }
     }
 
     pub fn update(&self, nice: Nice) {
-        self.weight.store(nice_to_weight(nice), Relaxed);
+        self.pending_weight
+            .store(nice_to_weight(nice), Ordering::Relaxed);
+        self.weight.fetch_or(HAS_PENDING, Ordering::Release);
     }
 
-    fn update_vruntime(&self, delta: u64) -> (u64, u64) {
-        let weight = self.weight.load(Relaxed);
+    fn update_vruntime(&self, delta: u64, weight: u64) -> u64 {
         let delta = delta * WEIGHT_0 / weight;
-        let vruntime = self.vruntime.fetch_add(delta, Relaxed) + delta;
-        (vruntime, weight)
+        self.vruntime.fetch_add(delta, Ordering::Relaxed) + delta
+    }
+
+    fn fetch_weight(&self) -> (u64, u64) {
+        let mut weight = self.weight.load(Ordering::Acquire);
+        if weight & HAS_PENDING == 0 {
+            return (weight, weight);
+        }
+
+        let mut new_weight = self.pending_weight.load(Ordering::Relaxed);
+        loop {
+            match self.weight.compare_exchange_weak(
+                weight,
+                new_weight,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            ) {
+                Ok(_) => break,
+                Err(failure) => {
+                    if failure & HAS_PENDING == 0 {
+                        return (failure, failure);
+                    }
+                    weight = failure;
+                    new_weight = self.pending_weight.load(Ordering::Relaxed);
+                }
+            }
+        }
+        let old_weight = weight & !HAS_PENDING;
+
+        let vruntime = self.vruntime.load(Ordering::Relaxed);
+        self.vruntime
+            .store(vruntime * old_weight / new_weight, Ordering::Relaxed);
+
+        (old_weight, new_weight)
     }
 }
 
@@ -226,12 +293,14 @@ impl SchedClassRq for FairClassRq {
             Some(EnqueueFlags::Spawn) => self.min_vruntime + self.vtime_slice(),
             _ => self.min_vruntime,
         };
+        let (_old_weight, weight) = fair_attr.fetch_weight();
+
         let vruntime = fair_attr
             .vruntime
-            .fetch_max(vruntime, Relaxed)
+            .fetch_max(vruntime, Ordering::Relaxed)
             .max(vruntime);
 
-        self.total_weight += fair_attr.weight.load(Relaxed);
+        self.total_weight += weight;
         self.entities.push(Reverse(FairQueueItem(entity, vruntime)));
     }
 
@@ -247,7 +316,12 @@ impl SchedClassRq for FairClassRq {
         let Reverse(FairQueueItem(entity, _)) = self.entities.pop()?;
 
         let sched_attr = entity.as_thread().unwrap().sched_attr();
-        self.total_weight -= sched_attr.fair.weight.load(Relaxed);
+        let (old_weight, _weight) = sched_attr.fair.fetch_weight();
+        // Equals to:
+        //
+        // self.total_weight = self.total_weight + weight - old_weight;
+        // self.total_weight -= weight;
+        self.total_weight -= old_weight;
 
         Some(entity)
     }
@@ -261,7 +335,11 @@ impl SchedClassRq for FairClassRq {
         match flags {
             UpdateFlags::Yield => true,
             UpdateFlags::Tick | UpdateFlags::Wait => {
-                let (vruntime, weight) = attr.fair.update_vruntime(rt.delta);
+                let (old_weight, weight) = attr.fair.fetch_weight();
+                if old_weight != weight {
+                    self.total_weight = self.total_weight + weight - old_weight;
+                }
+                let vruntime = attr.fair.update_vruntime(rt.delta, weight);
                 self.min_vruntime = match self.entities.peek() {
                     Some(Reverse(leftmost)) => vruntime.min(leftmost.key()),
                     None => vruntime,

--- a/kernel/src/sched/sched_class/mod.rs
+++ b/kernel/src/sched/sched_class/mod.rs
@@ -179,7 +179,17 @@ impl SchedAttr {
     }
 
     pub fn update_policy<T>(&self, f: impl FnOnce(&mut SchedPolicy) -> T) -> T {
-        self.policy.update(f)
+        self.policy.update(|policy| {
+            let ret = f(policy);
+            match *policy {
+                SchedPolicy::RealTime { rt_prio, rt_policy } => {
+                    self.real_time.update(rt_prio.get(), rt_policy);
+                }
+                SchedPolicy::Fair(nice) => self.fair.update(nice),
+                _ => {}
+            }
+            ret
+        })
     }
 
     fn last_cpu(&self) -> Option<CpuId> {

